### PR TITLE
adds disk caching.  Closes #470.

### PR DESCRIPTION
**Pull request by 345andrew**
In this PR:

* A successful load of data from the API will write to local disk if the useDiskCaching option is on.
* A failure to read from the API will read from the disk cache if the useDiskCaching option is on.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,6 +66,7 @@ log | The log provider. | Optional.  Defaults to a new bunyan.Logger with basic
 subscription | The name of your cloco subscription.  | Optional, but if not supplied will use the machine default.
 ttl | The time in seconds for items to live in the cache before refreshing. | Optional, if not supplied then cache items will never expire.
 url | The cloco API url. | Optional.  Will default to the public hosted cloco API.  Use this to override for on-premise installations of cloco.
+useDiskCaching | Indicates whether disk caching is used. | Optional.  Will default to false.  When set to true any information received from the API will be copied to disk, and will be loaded from disk on app start if the API cannot be contacted.
```
Comment (yaythomas) [inline README.md]: On the world's most minor list: copied to disk, rather than save?

File: src/cloco-client.spec.ts [hunk 2]
```diff
@@ -368,6 +688,38 @@ describe("ClocoClient unit tests", function(): void {
...
+          expect(dispatchSpy).toHaveBeenCalledTimes(1); // shows that the config published once.
+          done();
+        })
+        .catch((e: Error) => {
+          console.log(e);
+          fail("error not expected");
+          done();
+        });
+  });
+
```
Comment (yaythomas) [inline src/cloco-client.spec.ts:723]: have a ❤️  for all the tests

File: src/cloco-client.ts [hunk 3]
```diff
@@ -150,12 +151,27 @@ export class ClocoClient {
...
+      let filename: string = `~/.cloco/cache/application_${this.options.application}`;
 
       try {
         this.app = await ApiClient.getApplication(this.options);
+
+        // write to disk cache if in options.
+        if (this.options.useDiskCaching) {
+          Logger.log.debug(`ClocoClient.initializeApplication: writing application ${this.options.application} to disk.`);
+          let s: any = shell as any; // incorrect typing.
+          s.echo(JSON.stringify(this.app)).to(filename);
```
Comment (yaythomas) [inline src/cloco-client.ts]: why you dropping to shell here? node has a couple of ways to write to file without busting out a separate process. . . .

starting with the built-in https://nodejs.org/api/fs.html#fs_fs_writefile_file_data_options_callback

```
fs.writeFile('message.txt', 'Hello Node.js', (err) => {
  if (err) throw err;
  console.log('It\'s saved!');
});
```

File: src/cloco-client.ts [hunk 4]
```diff
@@ -150,12 +151,27 @@ export class ClocoClient {
     private async initializeApplication(): Promise<void> {
 
       Logger.log.debug(`ClocoClient.initializeApplication: start.`);
+      let filename: string = `~/.cloco/cache/application_${this.options.application}`;
```
Comment (yaythomas) [inline src/cloco-client.ts]: colour me lazy - what's in `this.options.application` - a guid or text? I'm thinking what happens if you have >1 node client running or something of the sort.

File: src/cloco-client.ts [hunk 5]
```diff
@@ -181,26 +197,49 @@ export class ClocoClient {
...
+        if (this.options.useDiskCaching) {
+          Logger.log.debug(`ClocoClient.loadConfigurationObjectWrapperFromApi: writing configuration '${objectId}' to disk.`);
+          let s: any = shell as any; // incorrect typing.
+          s.echo(JSON.stringify(wrapper)).to(filename);
+        }
+      } catch (e) {
+        this.onError.dispatch(this, e);
+        Logger.log.error(e, "ClocoClient.loadConfigurationObjectWrapperFromApi: Error encountered loading configuration from api.");
+        if (initializing) {
+          if (this.options.useDiskCaching && shell.test("-f", filename)) {
```
Comment (yaythomas) [inline src/cloco-client.ts]: ditto shell test for existence

**Review by yaythomas — COMMENTED** [review]
So it only gets from cache in case of error? I'd imagine for perf or something people might wanna retrieve from cache anyway, error or not?

File: src/cloco-client.ts (hunk 4 above)
Comment (345andrew) [reply src/cloco-client.ts]: The idea is that your cloco application maps 1:1 with your actual nodejs application.  Therefore you pass into the options the name of the subscription and the application and the client will load up all of the configuration for the application and keep it refreshed.

You could have >1 node client running on the same machine but I don't envisage *at the moment* having >1 running in the same nodejs application

File: src/cloco-client.ts (hunk 3 above)
Comment (345andrew) [reply src/cloco-client.ts]: Have gone through this and changed over to fs

File: src/cloco-client.ts (hunk 5 above)
Comment (345andrew) [reply src/cloco-client.ts]: ditto have changed over to fs

File: src/file-system.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+        Logger.log.debug(`FileSystem.writeFile: writing '${filename}' to disk.`);
+
+        return new Promise<void>(
+
+            function(
+                resolve: (value?: PromiseLike<void>) => void,
+                reject: (reason?: any) => void): void {
+
+                fs.writeFile(filename, data, "utf8", function(err: Error): void {
```
Comment (yaythomas) [inline src/file-system.ts:57]: this works assuming you're dealing with text - I've a suspicion binary bytes[] will not survive this?

I also have a suspicion this is bigger than a quick fix, so more flagging it for future TODOs.

File: src/settings.ts [hunk 7]
```diff
@@ -74,57 +74,55 @@ export class Settings {
    * @return {string} The bearer token.
    */
   public static getBearerToken(): string {
-    return Settings.readFileContent("~/.cloco/config/cloco_token");
+    return Settings.readFileContent(`${process.env.HOME}/.cloco/config/cloco_token`);
```
Comment (yaythomas) [inline src/settings.ts]: feel free to keep for a later TODO, but I've a suspicion this isn't going to x-platform - people can `npm install cloco-node-client` on any old O/S.  Had it been me I'd have made a local prop for `homeDir: string`, or maybe a `getHome(): string` and then used that throughout so it's nice and easy to change when the time comes, but I guess that could also be done when the time comes to worry about it.

**Review by yaythomas — APPROVED** [review]
:shipit:

File: src/settings.ts (hunk 7 above)
Comment (345andrew) [reply src/settings.ts]: That's a good idea actually, as I know that it's different on win32.  

function getUserHome() {
  return process.env[(process.platform == 'win32') ? 'USERPROFILE' : 'HOME'];
}

[http://stackoverflow.com/questions/9080085/node-js-find-home-directory-in-platform-agnostic-way](http://stackoverflow.com/questions/9080085/node-js-find-home-directory-in-platform-agnostic-way)

Comment (345andrew) [reply src/settings.ts]: Having said that, the cloco-bash tool only supports *nix

Comment (yaythomas) [reply src/settings.ts]: I thought MS joined proper IT and finally got themselves some bash? Presumably that means cloco-bash will work in that context?

**Pull request merged**